Extension/src/LanguageServer/cppBuildTaskProvider.ts

```diff
@@ -0,0 +1,357 @@
+                group: { kind: "build", "isDefault": true }
+            };
+            rawTasksJson.tasks.push(task);
+        }
+
+        // TODO: It's dangerous to overwrite this file. We could be wiping out comments.
+        const settings: OtherSettings = new OtherSettings();
+        const tasksJsonPath: string | undefined = getTasksJsonPath();
+        if (!tasksJsonPath) {
+            throw new Error("Failed to get tasksJsonPath in ensureBuildTaskExists()");
+        }
+
+        await util.writeFileText(tasksJsonPath, JSON.stringify(rawTasksJson, null, settings.editorTabSize));
+    }
+}
+
+class CustomBuildTaskTerminal implements vscode.Pseudoterminal {
+    private writeEmitter  = new vscode.EventEmitter<string>();
+    private closeEmitter = new vscode.EventEmitter<number>();
+    public get onDidWrite(): vscode.Event<string> { return this.writeEmitter.event; }
+    public get onDidClose(): vscode.Event<number> { return this.closeEmitter.event; }
+
+    private fileWatcher: vscode.FileSystemWatcher | undefined;
+
+
+    constructor(private command: string, private args: string[], private options: cp.ExecOptions | undefined, private workspaceRoot: string) {
+    }
+
+
+    open(_initialDimensions: vscode.TerminalDimensions | undefined): void {
+        telemetry.logLanguageServerEvent("cppBuildTaskStarted");
+        const pattern: string = path.join(this.workspaceRoot, 'cppBuild');
+        this.fileWatcher = vscode.workspace.createFileSystemWatcher(pattern);
+        this.fileWatcher.onDidChange(() => this.doBuild());
+        this.fileWatcher.onDidCreate(() => this.doBuild());
+        this.fileWatcher.onDidDelete(() => this.doBuild());
+        // At this point we can start using the terminal.
+        this.writeEmitter.fire("Starting build...\r\n");
+        this.doBuild();
+    }
+
+    close(): void {
+        // The terminal has been closed. Shutdown the build.
+        if (this.fileWatcher) {
+            this.fileWatcher.dispose();
+        }
+    }
+
+    private async doBuild(): Promise<number> {
+        return new Promise<number>((resolve, reject) => {
+            // Do build.
+            let activeCommand: string = util.resolveVariables(this.command, this.AdditionalEnvironment);
+            this.args.forEach(value => {
+                let temp: string = util.resolveVariables(value, this.AdditionalEnvironment);
+                if (temp && temp.includes(" ")) {
+                    temp = "\"" + temp + "\"";
+                }
+                activeCommand = activeCommand + " " + temp;
+            });
+            if (this.options?.cwd) {
+                this.options.cwd = util.resolveVariables(this.options.cwd, this.AdditionalEnvironment);
+            }
+            cp.exec(activeCommand, this.options, (_error, stdout, _stderr) => {
+                if (_error) {
+                    telemetry.logLanguageServerEvent("cppBuildTaskError", { "error": _error.message });
```

Comment:
I don't think we can safely log the "_error.message". This might have PII in it.  Let's just log the event name by itself instead.